authorize/ssh_grpc.go

```diff
@@ -140,26 +121,67 @@ func (a *Authorize) EvaluateSSH(ctx context.Context, streamID uint64, req ssh.Au
 
 	skipLogging := req.LogOnlyIfDenied && allowed
 	if !skipLogging {
-		s, _ := a.getDataBrokerSessionOrServiceAccount(ctx, req.SessionID, 0)
-
-		var u *user.User
-		if s != nil {
-			u, _ = a.getDataBrokerUser(ctx, s.GetUserId())
-		}
-		a.logAuthorizeCheck(ctx, &evalreq, res, s, u)
+		a.fetchSessionAndLogAuthorizeCheck(ctx, req.SessionID, evalreq, res)
 	}
 
 	return res, nil
 }
 
-func (a *Authorize) EvaluateUpstreamTunnel(_ context.Context, _ ssh.AuthRequest, _ *config.Policy) (*evaluator.Result, error) {
-	// XXX: temporary stub
-	return &evaluator.Result{
-		Allow: evaluator.NewRuleResult(true),
-		Deny:  evaluator.NewRuleResult(false),
-	}, nil
+func (a *Authorize) EvaluateUpstreamTunnel(ctx context.Context, req ssh.AuthRequest, route *config.Policy) (*evaluator.Result, error) {
+	ctx = a.withQuerierForCheckRequest(ctx)
+
+	evalreq := baseEvaluatorRequestFromSSHRequest(req)
+	evalreq.Policy = route
+
+	res, err := a.state.Load().evaluator.Evaluate(ctx, evalreq)
+	if err != nil {
+		log.Ctx(ctx).Error().Err(err).Msg("EvaluateUpstreamTunnel: error during OPA evaluation")
+		return nil, err
+	}
+
+	a.fetchSessionAndLogAuthorizeCheck(ctx, req.SessionID, evalreq, res)
```

Comment:
I think we should log these decisions differently (maybe at trace level, and in such a way that this isn't called at all if it is disabled). It could be hard to differentiate between these and downstream requests, and could be very noisy if there are a lot of routes. Failing an authorize check here is also not really a noteworthy event like it would be normally.